IA-2977 Refactoring some database fields to accommodate Azure VM

Qi77Qi opened the pull request:
Part of https://broadworkbench.atlassian.net/browse/IA-2977

- Mostly look at `20211015_support_proxy_to_azure_vm.xml` to see what's being refactored. The majority of the diff is due to renaming `googleProject` to `cloudContext`, which is term used workspace manager. It's also documented in [here](https://docs.google.com/document/d/1qkte2Y1SB6RR34mFhVNkpqb377Ta5y44yeJh6GUS43g/edit#heading=h.lsjdzdaj8yo7). Not all intended refactoring is done yet to keep the size of the PR.
- Added `proxy/az/v1/runtimes` route, but this is not anywhere close to working yet. This will be actively worked on tgt with more refactoring in a follow up PR


Verified the migration works against dev DB

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

Qi77Qi committed:
refactoring in prep for supporting azure

tmp

tmp

refactoring

bad merge

adjust path

remove googleId from PD table

PR feedback

fix things

add cloudProvider to both disk and runtime table

fix compile error

fix validation

address comments

Qi77Qi commented:
jenkins retest

Qi77Qi commented:
jenkins retest

jdcanas commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/models.scala:
```diff
@@ -32,3 +32,7 @@ object AppContext {
+
+final case class WorkspaceId(value: String) extends AnyVal
+
+final case class CloudContext(value: String) extends AnyVal
```
Almost feels like this should be slightly more intelligent than a simple value class. It seems a little pointless, if the intent is to wrap a GoogleProject AND an Azure string identifier, to represent them with more-or-less a string wrapper.

Namely, I think the data structure should express if its a googleProject or [meaningfulAzureIdentifierName] to the consumer at the call site.

Qi77Qi replied:
yea, I'm imagining we'll have such an ADT you're talking about in addition to this simple value wrapper...this simple value structure is still needed becuz we need to serielize/deserielize it with DB...from `CLUSTER` table, we can't really tell which context it is when reading from DB.

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala:
```diff
@@ -1,24 +1,22 @@
                                       zone: ZoneName,
                                       name: DiskName,
-                                      googleId: Option[GoogleId],
+                                      googleId: Option[ProxyHostName],
```
should this still be called googleId?

Qi77Qi replied:
probably not...there're probably still cases here and there that Type and name don't exactly match, but I think we can fix over time....I'll fix this one

Qi77Qi replied:
after a closer look..it seems like this `googleId` isn't really used...so I'm going to attempt to remove it

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/PersistentDiskComponent.scala:
```diff
@@ -137,15 +135,17 @@ object persistentDiskQuery {
-      .filter(_.googleProject === googleProject)
+      .filter(
+        _.cloudContext === cloudContext
+      ) //TODO: rename googleProject in PD table to CloudContext
```
this TODO seems done

jdcanas commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/DiskRoutes.scala:
```diff
@@ -194,9 +197,10 @@ object DiskRoutes {
+  implicit val getDiskResponseEncoder: Encoder[GetPersistentDiskResponse] = Encoder.forProduct13(
     "id",
     "googleProject",
+    "cloudContext",
```
I think there's a bug here based on how the signature of `GetPersistentDiskResponse` changes here (the number of fields don't seem to change in the linked signature but the for product number does?)

https://github.com/DataBiosphere/leonardo/pull/2339/files#diff-69288d2ed640980de319fedf8f45eb9f03c794e90cda4239dac24cb1ebe0e9d0R40

Qi77Qi replied:
yea...for backwards compatiblity, both `cloudContext` and `googleProject` are populated with `cloudContext` value in the actual response json...but the model only has one field, which is `cloudContext`...

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,57 @@
+        </column>
+    </addColumn>
+
+    <dropNotNullConstraint columnName="runtimeConfigId" tableName="CLUSTER" columnDataType="bigint(20)"/>
```
Curious why we need to drop the not-null and foreign key on `runtimeConfigId`?

Qi77Qi replied:
oh, so I initially wasn't planning on refactoring runtimeConfig table to account for Azure data as well...forgot to remove this after our convo last time

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,57 @@
+                  oldColumnName="clusterName"
+                  tableName="CLUSTER"/>
+
+    <renameColumn newColumnName="cloudContext"
```
There is an index on `(googleProject, runtimeName, destroyedDate)`… will the liquibase `renameColumn` update the index automatically w/ the new names?

Qi77Qi replied:
ah good call....didn't notice that..will check

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,57 @@
+
+    <dropForeignKeyConstraint baseTableName="CLUSTER" constraintName="FK_CLUSTER_RUNTIME_CONFIG_ID"/>
+
+    <createTable tableName="WSM_TO_CLUSTER">
```
Maybe `CLUSTER_WORKSPACE` would fit our naming pattern better?

Similar to:
```
MySQL [leonardo]> DESCRIBE CLUSTER_IMAGE;
+---------------+---------------+------+-----+----------------------+-------+
| Field         | Type          | Null | Key | Default              | Extra |
+---------------+---------------+------+-----+----------------------+-------+
| clusterId     | bigint(20)    | NO   | MUL | NULL                 |       |
| imageType     | varchar(254)  | YES  |     | NULL                 |       |
| imageUrl      | varchar(1024) | YES  |     | NULL                 |       |
| timestamp     | timestamp(6)  | NO   |     | CURRENT_TIMESTAMP(6) |       |
| homeDirectory | varchar(254)  | YES  |     | NULL                 |       |
+---------------+---------------+------+-----+----------------------+-------+
5 rows in set (0.037 sec)
```

rtitle replied:
Thinking more, maybe `RUNTIME_CONTROLLED_RESOURCE`?

Idk, that's what I put in this doc though: https://docs.google.com/document/d/1aZPZkzIqJ5P-HBSxcu65uwRPUaBRKodqdZ59W9ayw_I/edit#heading=h.i8n1rfgcgf3o

Qi77Qi replied:
yea think I like `RUNTIME_CONTROLLED_RESOURCE`

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog logicalFilePath="leonardo" xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www. ...
```
Would be good to test this on some real dev snapshots :)

Qi77Qi replied:
yea..I guess I can try running a local leo that points to a dev snapshots....lmk if you have a better way to do this

rtitle replied:
Yeah that's what I've done in the past too: cloned the dev DB here: https://console.cloud.google.com/sql/instances?project=broad-dsde-dev and pointed a Leo to it

Qi77Qi commented:
weird fiab error...jenkins retest

Qi77Qi commented:
```
20:31:48 [ERROR] Ran out of retries without successful config render for thurloe.
```

jenkins retest

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,75 @@
+        </column>
+    </addColumn>
+
+    <dropForeignKeyConstraint baseTableName="CLUSTER" constraintName="FK_CLUSTER_RUNTIME_CONFIG_ID"/>
```
Do we need to drop `FK_CLUSTER_RUNTIME_CONFIG_ID`?

Qi77Qi replied:
oops...this was leftoever of when I thought we need to make `runtimeConfigId` nullable...thanks for the catch!

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,75 @@
+                  oldColumnName="clusterName"
+                  tableName="CLUSTER"/>
+
+    <renameColumn newColumnName="cloudContext"
```
Have been thinking… the managed resource group by itself does not fully define the cloud context. The Azure cloud context is really the combination of `(mrg, subscription, tenant)`. Those 3 pieces of information are needed to make cloud calls in a user's cloud context.

By contrast for GCP only the `googleProject` is needed because the organization is implied (firecloud.org) and the billing account is contained within the googleProject.

So I think it's fine to have `cloudContext: varchar` in the DB tables for uniqueness checks… just noting that to get the _actual_ Azure cloud context (those 3 fields) Leo will have to make a [query](https://workspace.dsde-dev.broadinstitute.org/swagger-ui.html#/Workspace/getWorkspace) to WorkspaceManager.

Qi77Qi replied:
talked to Rob over slack...what we'll do for now is having `mrg,subscription,tenant` in the case of Azure...if this turns out to be not easy to work with down the line we can revisit this

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/models.scala:
```diff
@@ -32,3 +34,40 @@ object AppContext {
+  }
+  final case class Azure(value: ManagedResourceGroup) extends CloudContext {
+    override val asString = value.value
+    override val asStringWithProvider = s"AZURE/${value.value}"
```
Nit: maybe `Azure` capitalization if this string will be user-facing.

Qi77Qi replied:
this string is mostly used in exception messages...so it might bubble up to user...

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,75 @@
+        <column name="runtimeId"/>
+    </createIndex>
+    <addForeignKeyConstraint baseColumnNames="runtimeId" baseTableName="RUNTIME_CONTROLLED_RESOURCE" constraintName="FK_RUNTIME_CONTROLLED_RESOURCE" deferrable="false" initiallyDeferred="false" onDelete="NO ACTION" onUpdate="NO ACTION" referencedColumnNames="id" referencedTableName="CLUSTER"/>
+    <addUniqueConstraint columnNames="runtimeId, workspaceId, resourceId" constraintName="IDX_RUNTIME_CONTROLLED_RESOURCE_UNIQUE" tableName="RUNTIME_CONTROLLED_RESOURCE"/>
```
Should this be unique on `(runtimeId, workspaceId, resourceId, resourceType)`?

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/package.scala:
```diff
@@ -22,7 +22,9 @@ package object api {
     Marshaller(implicit ec => (x => m(x.unsafeToFuture()(cats.effect.unsafe.IORuntime.global))))
 
   val googleProjectSegment = Segment.map(GoogleProject)
+  val managedResourceGroupSegment = Segment.map(ManagedResourceGroup)
```
Not sure we need `managedResourceGroupSegment` -- were we going to put `workspaceId` in the v2 routes?

Qi77Qi replied:
ah yea..forgot to update this after that discussion

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/api/ProxyRoutes.scala:
```diff
@@ -52,6 +51,18 @@ class ProxyRoutes(proxyService: ProxyService, corsSupport: CorsSupport, refererC
                       }
                   }
+                } ~ pathPrefix("v2" / "runtimes" / "azure") {
+                  pathPrefix(managedResourceGroupSegment / runtimeNameSegment) { (mrg, runtimeName) =>
```
I didn't fully update the [doc](https://docs.google.com/document/d/1aZPZkzIqJ5P-HBSxcu65uwRPUaBRKodqdZ59W9ayw_I/edit#), but perhaps we want:

```
/proxy/v2/runtimes/{workspaceId}/gcp/{name}
/proxy/v2/runtimes/{workspaceId}/azure/{name}
```
similar to runtimes routes.

Or maybe leave out proxy route changes from this PR since it's descoped from the demo?

Qi77Qi replied:
I updated the route to `/proxy/v2/runtimes/{workspaceId}/azure/{name}`, but left it as not implemented

rtitle reviewed (commented):
The refactoring looks good.. couple minor comments.

One comment above that `managedResourceGroup: String` does not fully represent the Azure cloud context (you need tenant and subscription too). However not sure if Leo really needs to store this information since it can look it up from WSM when needed. Anyway I think this is a good first pass and we can iterate if we need to change more fields once we start adding the new routes/WSM calls.

kyuksel commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,75 @@
+<databaseChangeLog logicalFilePath="leonardo" xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd http://www.liquibase.org/xml/ns/dbchangelog http://www. ...
+
+<changeSet logicalFilePath="leonardo" author="qi" id="support_proxy_to_azure_vm">
+    <addColumn tableName="CLUSTER">
```
Would it be good to rename the table `CLUSTER` since everywhere else related we've shifted to using the term `runtime`?

Qi77Qi replied:
definitely..probably not for this PR tho since I'm just trying to wrap this up at this point... we won't continue this work until later after we feel good about demo

Qi77Qi committed:
bad merge

jdcanas reviewed (approved):
Looks good as a first pass at a leo cloud context. I imagine the 'cloudResource' field may need to be a more complex object in the future, otherwise i'd call it 'cloudResourceName'.  

Seems like one of those 'trust the tests' changes. Might run a multi-test too.

Qi77Qi commented:
@jdcanas yea, we'll start with `mrg,subscription,tenant` for `cloudContext` field for azure. Altho there're also some conversations about relationships between workspaces and subscriptions etc, which might change. But I think having it as `mrg,subscription,tenant` is a good first pass. This will be the combo we need for retrieving an azure resource

Qi77Qi committed:
bad merge

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/JsonCodec.scala:
```diff
@@ -275,6 +286,18 @@ object JsonCodec {
+        case CloudProvider.Gcp =>
+          x.downField("cloudResource").as[GoogleProject].map(p => CloudContext.Gcp(p))
+        case CloudProvider.Azure =>
+          x.downField("cloudResource").as[ManagedResourceGroup].map(p => CloudContext.Azure(p))
```
Are we still thinking of embedding `"tenantId,subscriptionId,managedResourceGroupName"` as the Azure cloud context? I'm wondering If the decoder should split that string into separate fields..

Qi77Qi replied:
ah yea...forgot this needs to be refactored for that

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/models.scala:
```diff
@@ -0,0 +1,41 @@
+    override val asStringWithProvider = s"Gcp/${value.value}"
+    override def cloudProvider: CloudProvider = CloudProvider.Gcp
+  }
+  final case class Azure(value: ManagedResourceGroup) extends CloudContext {
```
Similar to above comment, wondering if `CloudContext.Azure` should have `(tenantId: TenantId, subscriptionId: SubscriptionId, mrg: ManagedResourceGroupName)`

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/azureModels.scala:
```diff
@@ -0,0 +1,3 @@
+package org.broadinstitute.dsde.workbench.leonardo
+
+final case class ManagedResourceGroup(value: String) extends AnyVal
```
Maybe `ManagedResourceGroupName`?

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala:
```diff
@@ -64,7 +64,7 @@ object Runtime {
 
     new URL(
-      urlBase + googleProject.value + "/" + runtimeName.asString + "/" + tool.proxySegment
+      urlBase + cloudContext.asString + "/" + runtimeName.asString + "/" + tool.proxySegment
```
I think this only works for the Google case, right? Just wondering if it should fail if it's an Azure cloud context for now.

Qi77Qi replied:
will do

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,74 @@
+        </column>
+    </addColumn>
+
+    <addColumn tableName="PERSISTENT_DISK">
```
Maybe would be nice to break up updates to different tables to different change sets? Just so if 1 fails it doesn't leave a table partially updated..

Qi77Qi replied:
I separated PD update and creating the new wsm resource table into a separate xml file.

rtitle commented on http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20211015_support_proxy_to_azure_vm.xml:
```diff
@@ -0,0 +1,74 @@
+                  tableName="PERSISTENT_DISK"/>
+
+    <addUniqueConstraint
+            columnNames="cloudContext,runtimeName,destroyedDate"
```
Should this be `cloudContext,cloudProvider,runtimeName,destroyedDate`?

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:
```diff
@@ -75,8 +77,6 @@ class ClusterTable(tag: Tag) extends Table[ClusterRecord](tag, "CLUSTER") {
   def customClusterEnvironmentVariables = column[Option[Map[String, String]]]("customClusterEnvironmentVariables")
   def deletedFrom = column[Option[String]]("deletedFrom")
 
-  def uniqueKey = index("IDX_CLUSTER_UNIQUE", (googleProject, clusterName, destroyedDate), unique = true)
```
Can we keep this `def`? I'm not sure what it actually does though..

Qi77Qi replied:
based on slick doc in source code, I think this just adds the unique index, which is duplicated by [this](https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/resources/org/broadinstitute/dsde/workbench/leonardo/liquibase/changesets/20171010_cluster_unique_index.xml#L5)....

rtitle commented:
The refactoring looks good overall. Main comment is whether we can model `CloudContext.Azure` as containing tenantId/subscriptionId/managedResourceGroupName. We will need those 3 fields in order to make Azure cloud calls using the Java client.

Qi77Qi committed:
comments and bad merge fixes

Qi77Qi commented:
@rtitle  thanks for re-review...addressed comments in [this](https://github.com/DataBiosphere/leonardo/pull/2339/commits/fbd13eb8fabc5c0a83d9b8b2f5036ead35cbc471)

rtitle reviewed (approved):
Thanks for addressing.. lgtm if tests pass

Qi77Qi commented:
```
message: Failed to create runtime terra-quality-cf3a5b27/automation-test-afy8ukx7z, status: 500 Internal Server Error body: {"source":"leonardo","message":"HTTP connection closed: https://sam-fiab.dsde-qa.broadinstitute.org:443","statusCode":500,"exceptionClass":"class java.net.SocketException","traceId":null}
```

jenkins retest

Qi77Qi commented:
multi-test

Qi77Qi commented:
multi-test

Qi77Qi commented:
one failure from multi-test that saw happened in dev branch today as well...

multi-test

Qi77Qi commented:
multi-test

Qi77Qi commented:
multi-test
